pp.c

```diff
@@ -7939,10 +7934,158 @@ PP(pp_argcheck)
+    AV  *defav         = GvAV(PL_defgv); /* @_ */
+
+    assert(!SvMAGICAL(defav));
+    UV argc = (UV)(AvFILLp(defav) + 1);
+
+    S_check_argc(aTHX_ argc, aux->params, aux->opt_params, aux->slurpy);
 
     return NORMAL;
 }
 
+PP(pp_multiparam)
+{
+    struct op_multiparam_aux *aux = (struct op_multiparam_aux *)cUNOP_AUX->op_aux;
+    UV nparams = aux->n_positional;
+    char slurpy = aux->slurpy;
+    AV  *defav = GvAV(PL_defgv); /* @_ */
+
+    assert(!SvMAGICAL(defav));
+    UV argc = (UV)(AvFILLp(defav) + 1);
+
+    S_check_argc(aTHX_ argc, nparams, nparams - aux->min_args, slurpy);
+
+    UV parami;
+    for(parami = 0; parami < nparams; parami++) {
+        PADOFFSET padix = aux->param_padix[parami];
+        if(!padix) {
+            if(argc)
+                argc--;
+            continue;
+        }
+
+        SV **padentry = &PAD_SVl(padix);
+        save_clearsv(padentry);
+
+        if(!argc) {
+            /* Ran out of arg values for this param. It must be a missing
+             * optional. Remark that it's missing so a subsequent OP_PARAMTEST
+             * knows */
+            SvPADSTALE_on(*padentry);
+            continue;
+        }
+
+        SV **valp = av_fetch(defav, parami, FALSE);
+        SV *val = valp ? *valp : &PL_sv_undef;
+        argc--;
+
+        assert(TAINTING_get || !TAINT_get);
+        if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+            TAINT_NOT;
+
+        SvSetMagicSV(*padentry, val);
+    }
+
+    if(!slurpy || !aux->slurpy_padix)
+        return PL_op->op_next;
+
+    /* Now we know we have a slurpy */
+    SV **padentry = &PAD_SVl(aux->slurpy_padix);
+    save_clearsv(padentry);
+
+    if(slurpy == '@') {
+        AV *av = (AV *)*padentry;
+        assert(SvTYPE(av) == SVt_PVAV);
+
+        if(av_count(av)) {
+            /* see "target should be empty" comments in pp_argelem above */
+            av_refresh_elements_range(defav, parami, parami + argc);
+            av_clear(av);
+        }
+
+        av_extend(av, argc);
+
+        IV avidx = 0;
+        for(; argc; parami++, argc--) {
+            SV **valp = av_fetch(defav, parami, FALSE);
+            SV *val = valp ? *valp : &PL_sv_undef;
+
+            assert(TAINTING_get || !TAINT_get);
+            if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                TAINT_NOT;
+
+            av_store(av, avidx++, newSVsv(val));
+        }
+    }
+    else if(slurpy == '%') {
+        HV *hv = (HV *)*padentry;
+        assert(SvTYPE(hv) == SVt_PVHV);
+
+        if(SvRMAGICAL(hv) || HvUSEDKEYS(hv) > 0) {
+            /* see "target should be empty" comments in pp_argelem above */
+            av_refresh_elements_range(defav, parami, parami + argc);
+            hv_clear(hv);
+        }
+
+        assert((argc % 2) == 0);
+
+        while(argc) {
+            SV **svp;
+
+            svp = av_fetch(defav, parami, FALSE); parami++;
+            SV *key = svp ? *svp : &PL_sv_undef;
```

Comment:
What on earth does `hv_store_ent(hv, &PL_sv_undef, newSVsv(val), 0);`  mean?

how about executing `hv_store_ent(hv, &PL_sv_undef, newSVsv(val), 0);` 5  or 9 times in a row, each time with a different addr in var "val"? what does that mean? why is the previous sentence valid correct behavior and not a panic/heap corruption?

Reply:
A SEGV here would be just as good and more efficient than an actual `croak()` fn call located here and its C string that nobody will ever see unless they use a disassembler.